external/builder/babel-plugin-pdfjs-preprocessor.mjs

```diff
@@ -0,0 +1,241 @@
+          jsonPath = joinPaths(
+            ctx.rootPath,
+            jsonPath.substring(ROOT_PREFIX.length)
+          );
+        }
+        return JSON.parse(fs.readFileSync(jsonPath, "utf8"));
+    }
+    throw new Error("Unsupported action");
+  } catch (e) {
+    throw path.buildCodeFrameError(
+      "Could not process " +
+        PDFJS_PREPROCESSOR_NAME +
+        "." +
+        actionName +
+        ": " +
+        e.message
+    );
+  }
+}
+
+function babelPluginPDFJSPreprocessor(babel, ctx) {
+  return {
+    name: "babel-plugin-pdfjs-preprocessor",
+    manipulateOptions({ parserOpts }) {
+      parserOpts.attachComment = false;
+    },
+    visitor: {
+      "ExportNamedDeclaration|ImportDeclaration": ({ node }) => {
+        if (node.source && ctx.map && ctx.map[node.source.value]) {
+          const newValue = ctx.map[node.source.value];
+          node.source.value = newValue;
+        }
+      },
+      "IfStatement|ConditionalExpression": {
+        exit(path) {
+          const { node } = path;
+          if (t.isBooleanLiteral(node.test)) {
+            // if (true) stmt1; => stmt1
+            // if (false) stmt1; else stmt2; => stmt2
+            path.replaceWith(
+              node.test.value === true
+                ? node.consequent
+                : node.alternate || t.emptyStatement()
+            );
+          }
+        },
+      },
+      UnaryExpression(path) {
+        const { node } = path;
+        if (node.operator === "typeof" && isPDFJSPreprocessor(node.argument)) {
+          // typeof PDFJSDev => 'object'
+          path.replaceWith(t.stringLiteral("object"));
+          return;
+        }
+        if (node.operator === "!" && t.isBooleanLiteral(node.argument)) {
+          // !true => false,  !false => true
+          path.replaceWith(t.booleanLiteral(!node.argument.value));
+        }
+      },
+      LogicalExpression: {
+        exit(path) {
+          const { node } = path;
+          if (!t.isBooleanLiteral(node.left)) {
+            return;
+          }
+
+          switch (node.operator) {
+            case "&&":
+              // true && expr => expr
+              // false && expr => false
+              path.replaceWith(
+                node.left.value === true ? node.right : node.left
+              );
+              break;
+            case "||":
+              // true || expr => true
+              // false || expr => expr
+              path.replaceWith(
+                node.left.value === true ? node.left : node.right
+              );
+              break;
+          }
+        },
+      },
+      BinaryExpression: {
+        exit(path) {
+          const { node } = path;
+          switch (node.operator) {
+            case "==":
+            case "===":
+            case "!=":
+            case "!==":
+              // folding == and != check that can be statically evaluated
+              const { confident, value } = path.evaluate();
```

Comment:
How "confident", pun intended, can we be about this *always* doing the right thing?

Some quick searching brought up https://github.com/babel/babel/issues/14197, which suggests that there *may* be problems in some cases (unless I'm misunderstanding how this works).